plugin/src/main/java/com/microsoft/java/bs/gradle/plugin/SourceSetsModelBuilder.java

```diff
@@ -42,198 +44,209 @@
+        Task compileTask = languageModelBuilder.getLanguageCompileTask(project, sourceSet);
+        if (compileTask != null) {
+          String compileTaskName = getFullTaskName(projectPath, compileTask.getName());
+          taskNames.add(compileTaskName);
         }
-        String displayName = projectName + " [" + gradleSourceSet.getSourceSetName() + ']';
-        gradleSourceSet.setDisplayName(displayName);
+        srcDirs.addAll(languageModelBuilder.getSourceFoldersFor(project, sourceSet));
+        generatedSrcDirs.addAll(
+            languageModelBuilder.getGeneratedSourceFoldersFor(project, sourceSet));
 
-        // source
-        Set<File> srcDirs = new HashSet<>();
-        Set<File> generatedSrcDirs = new HashSet<>();
-        for (LanguageModelBuilder languageModelBuilder :
-            GradleBuildServerPlugin.SUPPORTED_LANGUAGE_BUILDERS) {
-          if (languageModelBuilder.appliesFor(project, sourceSet)) {
-            Task compileTask = languageModelBuilder.getLanguageCompileTask(project, sourceSet);
-            if (compileTask != null) {
-              String compileTaskName = getFullTaskName(projectPath, compileTask.getName());
-              taskNames.add(compileTaskName);
-            }
-            srcDirs.addAll(languageModelBuilder.getSourceFoldersFor(project, sourceSet));
-            generatedSrcDirs.addAll(
-                languageModelBuilder.getGeneratedSourceFoldersFor(project, sourceSet));
+        LanguageExtension extension = languageModelBuilder.getExtensionsFor(project, sourceSet,
+            gradleSourceSet.getModuleDependencies());
+        if (extension != null) {
+          extensions.put(languageModelBuilder.getLanguageId(), extension);
+          if (extension.isJavaExtension()) {
+            sourceOutputDirs.add(extension.getAsJavaExtension().getClassesDir());
+          } else if (extension.isScalaExtension()) {
+            sourceOutputDirs.add(extension.getAsScalaExtension().getClassesDir());
           }
         }
-        gradleSourceSet.setSourceDirs(srcDirs);
-        exclusionFromDependencies.addAll(srcDirs);
-        gradleSourceSet.setGeneratedSourceDirs(generatedSrcDirs);
-        exclusionFromDependencies.addAll(generatedSrcDirs);
-
-        // classpath
-        List<File> compileClasspath = new LinkedList<>(sourceSet.getCompileClasspath().getFiles());
-        gradleSourceSet.setCompileClasspath(compileClasspath);
+      }
+    }
+    gradleSourceSet.setSourceDirs(srcDirs);
+    gradleSourceSet.setGeneratedSourceDirs(generatedSrcDirs);
+    gradleSourceSet.setExtensions(extensions);
+    gradleSourceSet.setSourceOutputDirs(sourceOutputDirs);
 
-        sourceSetsToClasspath.put(gradleSourceSet, compileClasspath);
+    // classpath
+    try {
+      Set<File> files = sourceSet.getCompileClasspath().getFiles();
+      List<File> compileClasspath = new LinkedList<>(files);
+      gradleSourceSet.setCompileClasspath(compileClasspath);
+    } catch (ResolveException ex) {
```

Comment:
Use `GradleException`? Since it may throw `org.gradle.internal.resolve.ModuleVersionNotFoundException` as well

https://github.com/microsoft/build-server-for-gradle/pull/122#issuecomment-2031099285

Reply:
I think it varies depending on the Gradle version you use.  I added the `missing-repository` test and found that `ResolveException` is thrown in `SourceSetsModelBuilder` for some versions and `ModuleVersionNotFoundException` is thrown in `DependencyCollector` in others.  So I've added `GradleException` to the `DependencyCollector` code.
I can change `ResolveException` to `GradleException` here if you want but I can't reproduce that.

Reply:
I'd like to use GradleException. Missing repository is one case to make the invocation failed. But there must be some other cases that we don't know which will fail the invocation too. Using GradleException makes me feel more confident

Reply:
Done